src/Data/LLVM/BitCode/GetBits.hs

```diff
@@ -11,68 +11,106 @@ module Data.LLVM.BitCode.GetBits (
   , skip
   ) where
 
-import Data.LLVM.BitCode.BitString
+import           Data.LLVM.BitCode.BitString
 
-import Control.Applicative (Alternative(..))
-import Control.Arrow (first)
-import Control.Monad (MonadPlus(..),when,replicateM_)
-import qualified Data.Binary.Get as BG
-import Data.Bits (shiftR)
-import Data.ByteString (ByteString)
-import Data.Word (Word32)
+import           Control.Applicative ( Alternative(..) )
+import           Control.Monad ( MonadPlus(..) )
+import           Data.Bits ( (.&.), (.|.), shiftL, shiftR )
+import           Data.ByteString ( ByteString )
+import qualified Data.ByteString as BS
 
 #if !MIN_VERSION_base(4,13,0)
-import Control.Monad.Fail( MonadFail )
+import           Control.Monad.Fail ( MonadFail )
 import qualified Control.Monad.Fail
 #endif
 
 -- Bit-level Parsing -----------------------------------------------------------
 
-newtype GetBits a = GetBits { unGetBits :: SubWord -> BG.Get (a,SubWord) }
+newtype GetBits a = GetBits { unGetBits :: Either String (BitPosition -> BS.ByteString -> (BitsGetter a, BitPosition)) } -- Left is outer fail
+
+newtype BitPosition = BitPosition (NumBits, NumBits) -- (current, limit)
+
+type BitsGetter a = BS.ByteString -> Either String a -- Left is inner fail
+
 
 -- | Run a @GetBits@ action, returning its value, and the number of bits offset
 -- into the next byte of the stream.
-runGetBits :: GetBits a -> BG.Get a
-runGetBits m = fst `fmap` unGetBits m aligned
+runGetBits :: GetBits a -> ByteString -> Either String a
+runGetBits m bs =
+  let startPos = BitPosition ( Bits' 0
+                             , bytesToBits $ Bytes' $ BS.length bs
+                             )
+  in ($ bs) =<< (fst . ($ bs) . ($ startPos) <$> unGetBits m)
+
 
 instance Functor GetBits where
   {-# INLINE fmap #-}
-  fmap f m = GetBits (\ off -> first f <$> unGetBits m off)
+  fmap f m = GetBits $ case unGetBits m of
+    Left e -> Left e
+    Right g -> Right $ \pos inp -> let (b,n) = g pos inp
+                                   in (\bs -> f <$> b bs, n)
 
 instance Applicative GetBits where
   {-# INLINE pure #-}
-  pure x = GetBits (\ off -> return (x,off))
+  pure x = GetBits $ pure $ \pos _ -> (const $ pure x, pos)
 
   {-# INLINE (<*>) #-}
-  f <*> x = GetBits $ \ off0 -> do
-    (g,off1) <- unGetBits f off0
-    (y,off2) <- unGetBits x off1
-    return (g y,off2)
+  f <*> x =
+    GetBits $ do k <- unGetBits f
+                 l <- unGetBits x
+                 return $ \pos inp -> let (g,n) = k pos inp
+                                          (y,m) = l n inp
+                                          in ( \_ -> do g' <- g inp
+                                                        y' <- y inp
+                                                        return $ g' y'
```

Comment:
Perhaps I'm not getting what the argument to `BitsGetter` represents, but I'm surprised to see this `BitsGetter` completely throw away its `ByteString` argument. What is the intuition behind this?

Reply:
What you are seeing here is actually an awkward window into the evolution of this code.  The initial approach was attempting to defer the need for the ByteString to the inner portion (i.e. I wanted `newtype GetBits a = GetBits { unGetBits :: Either String (BitPosition -> (BitsGetter a, BitPosition) ) }` instead of the implementation in this PR) but in order to implement `Alternative` `<|>`, it was necessary to have the `ByteString` as an argument to `unGetBits` itself.  Thus, the `\_` on line 63 is the same value as `inp` (introduced at the outer layer on line 61).  This PR does not yet contain the refactoring to remove these duplicate inner arguments (that is coming in one of the future PR's).

As noted in a side-band communication but worth explicitly documenting here in this context: this is part of a series of PR's that iterate through a set of work designed to improve performance and reduce memory consumption.  They are being introduced as individual PR's to capture the different phases of work and provide PR-level higher level information about the general intent of those phases.  This leads to the unfortunately frequent use of "this will be changed in a future PR"; I apologize for the frequency of these uses and the corresponding request for deferment of judgement, but I think this process is *valuable* because it's easier to review smaller steps than a single massive PR, and I am committed to returning to these individual comments made during the process once we reach the endpoint (wherever that is) and re-address them.